gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorage.java

```diff
@@ -200,6 +203,43 @@ void copy(
       List<String> dstObjectNames)
       throws IOException;
 
+  /**
+   * Copies metadata of the given objects. After the copy is successfully complete, each object blob
+   * is reachable by two different names. Copying between two different locations or between two
+   * different storage classes is not allowed.
+   *
+   * @param srcObjects list of the objects to copy
+   * @param dstObjects list of the objects after copy
+   * @throws java.io.FileNotFoundException if the source object or the destination bucket does not
+   *     exist
+   * @throws IOException in all other error cases
+   */
+  default void copy(List<StorageResourceId> srcObjects, List<StorageResourceId> dstObjects)
+      throws IOException {
+    checkArgument(srcObjects != null, "srcObjects must not be null");
+    checkArgument(dstObjects != null, "dstObjects must not be null");
+
+    List<String> srcObjectNames =
+        srcObjects.stream().map(StorageResourceId::getObjectName).collect(Collectors.toList());
+    List<String> dstObjectNames =
+        dstObjects.stream().map(StorageResourceId::getObjectName).collect(Collectors.toList());
+    String srcBucketName = srcObjects.get(0).getBucketName();
+    String dstBucketName = dstObjects.get(0).getBucketName();
+
+    for (int i = 0; i < srcObjects.size(); i++) {
+      if (!srcBucketName.equals(srcObjects.get(i).getBucketName()))
+        throw new UnsupportedOperationException(
```

Comment:
Hm, after further thinking, if GCS does not have this limitation, why should we impose it artificially?

Reply:
For the default Impl, I think we should impose this check, to ensure that compatibility with existing `copy` method.

If we are ok, breaking compatibility, we can remove default impl, and the check as well